CAIPs/caip-294.md

```diff
@@ -0,0 +1,431 @@
+  `caip294:${selected_uuid}`,
+  {
+    detail: sessionRequest
+  }
+));
+
+
+
+// 4. After the response was received by the Blockchain Library from the wallet
+// provider then the session is established with a sessionId matchin the UUID
+// thus signing requests can be using a CAIP-27 request to the wallet user
+const signingRequest = {
+  id: 456,
+  jsonrpc: "2.0",
+  method: "wallet_requestMethod",
+  params: {
+    // UUID from WalletData is used as SessionId
+    sessionId: walletData.uuid,
+    scope: "chain:777",
+    request: {
+      method: "chain_signMessage",
+      params: [
+        "Verifying my wallet with this message",
+        "0xa89Df33a6f26c29ea23A9Ff582E865C03132b140",
+      ],
+    },
+  },
+};
+
+let signingResult = {}
+
+window.addEventListener(`caip294:${selected_uuid}`, (event) => {
+  if (event.detail.id === signingRequest.id) {
+    // Get JSON-RPC response
+    if (event.detail.error) {
+      console.error(event.detail.error.message);
+    } else {
+      signingResult = event.detail.result
+    }
+  }
+});
+
+window.dispatchEvent(new CustomEvent(
+  `caip294:${selected_uuid}`,
+  {
+    detail: signingRequest
+  }
+));
+```
+
+Logic from the Wallet Provider:
+
+```typescript
+// 1. Wallet Provider sets their WalletData and then listens to wallet_prompt message
+// and also immediatelly posts a message with the WalletData as wallet_announce type
+const walletData = {
+  uuid: generateUUID(); // eg. "350670db-19fa-4704-a166-e52e178b59d2"
+  name: "Example Wallet",
+  icon: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAHElEQVQI12P4//8/w38GIAXDIBKE0DHxgljNBAAO9TXL0Y4OHwAAAABJRU5ErkJggg==",
+  rdns: "com.example.wallet",
+}
+
+window.dispatchEvent(new CustomEvent(
+  "caip294:wallet_announce",
+  {
+    detail: {
+      id: 1,
+      jsonrpc: "2.0"
+      method: "wallet_announce",
+      params: walletData,
+    }
+  }
+));
+
+window.addEventListener("caip294:wallet_prompt", (event) => {
+  // when a prompt message was received then the wallet will announces again
+  window.dispatchEvent(new CustomEvent(
+    "caip294:wallet_announce",
+    {
+      detail: {
+        id: 1,
+        jsonrpc: "2.0"
+        method: "wallet_announce",
+        params: walletData,
+      }
+    }
+  ));
+});
+
+// 2. User presses "Connect Wallet" on the application webpage which will select UUID
+
+const selected_uuid = "350670db-19fa-4704-a166-e52e178b59d2"
+
+// 3. Wallet Provider receives a CAIP-25 request to establish a wallet connection
+// prompts the user to approve and once its approved it can respond back to app
+// Wallet Provider listens for request
+let sessionRequest = {}
+
+window.addEventListener(`caip294:${selected_uuid}`, (event) => {
+  if (event.detail.method === "wallet_createSession") {
+    sessionRequest = event.detail
+    // if incoming requests match the WalletData UUID
+    if (checkSupportedScopes(event.detail.params)) {
+        // prompt user to approve session
+    }
+  }
+});
```

Comment:
Could a different wallet 'hijack' these events? Eg if I'm a wallet, can I send a `wallet_prompt` message in the app, capture the response, find the current UUID of Metamask, and then listen to its events to eg. also display my own popup?